packages/editor/src/components/inserter/menu.js

```diff
@@ -345,21 +345,30 @@ export class InserterMenu extends Component {
 }
 
 export default compose(
-	withSelect( ( select, { rootClientId } ) => {
+	withSelect( ( select, { clientId, isAppender, rootClientId } ) => {
 		const {
 			getInserterItems,
 			getBlockName,
+			getBlockRootClientId,
+			getBlockSelectionEnd,
 		} = select( 'core/editor' );
 		const {
 			getChildBlockNames,
 		} = select( 'core/blocks' );
 
-		const rootBlockName = getBlockName( rootClientId );
+		let rootClientIdToUse = rootClientId;
+		if ( ! rootClientIdToUse && ! clientId && ! isAppender ) {
+			const end = getBlockSelectionEnd();
+			if ( end ) {
+				rootClientIdToUse = getBlockRootClientId( end ) || undefined;
+			}
+		}
+		const rootBlockName = getBlockName( rootClientIdToUse );
 
 		return {
 			rootChildBlocks: getChildBlockNames( rootBlockName ),
-			items: getInserterItems( rootClientId ),
-			rootClientId,
+			items: getInserterItems( rootClientIdToUse ),
+			rootClientId: rootClientIdToUse,
```

Comment:
I'm a little bit concerned about changing the `rootClientId` to be passed here.

`rootClientId` has a specific meaning as a prop of the `Inserter`. 

It means the inserter in which this inserter is rendered, it doesn't mean the inserter in which the block to be added is going to be added. If this second option is what we want, we should do the correction in `inserter/index.js` instead to ensure we have a coherent value across all the inserter components.

Also, I wonder if it should be a separate prop `rootClientId` is the external prop. `destinationRootClientId` is the prop used here.

Reply:
Hi @youknowriad I followed your suggestion, and now we pass destinationRootClientId as a distinct prop.
The end 2 end test is available at https://github.com/WordPress/gutenberg/pull/14054. Currently, it fails because this PR was not yet merged.